src/fairseq2/nn/utils/module.py

```diff
@@ -389,6 +390,59 @@ def infer_device(
     )
 
 
+def broadcast_module(
+    module: Module, gang: Gang, *, source_rank: int = 0, broadcast_buffers: bool = True
+) -> None:
+    """Broadcast ``module`` to all processes in ``gang``.
+
+    :param module:
+        The module to broadcast.
+    :param gang
+        The gang over which to broadcast ``module``.
+    :param source_rank:
+        The rank of the source process from which to broadcast.
+    :param broadcast_buffers:
+        If ``True``, broadcasts not only the parameters, but the buffers as well.
+    """
+    to_device(module, gang.device)
+
+    if gang.size == 1:
+        return
+
+    memo: Set[Tensor] = set()
+
+    tensors = []
+
+    for param in module.parameters():
+        if param in memo:
+            continue
+
+        memo.add(param)
+
+        tensors.append(param.detach())
```

Comment:
should this also support broadcasting the gradients for the tensors ?

Reply:
Broadcasting grads is trickier since if the `module` on the receiving side does not have grads (which would be the common case), than you also have to find a way to copy over the autograd graph and create grad tensors on the fly instead of doing just an all-reduce call. For now, we can log a warning if we see that the source module has grads, telling to the user that they won't be synced. If we really find a legitimate use case for syncing grads in the future, we can reconsider extending this function.

Reply:
#594 594

Reply:
Sounds good :)